posts/inside-rust/2021-03-05-async-vision-doc.md

```diff
@@ -0,0 +1,91 @@
+[dt]: https://rust-lang.github.io/wg-async-foundations/vision/tenets.html
+[cc]: https://rust-lang.github.io/wg-async-foundations/vision/characters.html
+[dd]: https://rust-lang.github.io/wg-async-foundations/design_docs.html
+[gf]: https://forms.gle/YKNniGhaNXBhmjXNA
+[gr]: https://is.gd/T6TadC
+[z]: https://rust-lang.zulipchat.com/
+[wgz]: https://rust-lang.zulipchat.com/#streams/187312/wg-async-foundations
+[amb]: https://rust-lang.github.io/wg-async-foundations/vision.html#think-big----too-big-if-you-have-to
+[cok]: https://en.wikipedia.org/wiki/Curse_of_knowledge
+[workarounds]: https://rust-lang.github.io/async-book/07_workarounds/01_chapter.html
+
+The [Async Foundations Working Group][wg] believes Rust can become one of the most popular choices for building network applications and services. People are already building all kinds of networked applications using Async Rust, ranging from embedded devices to clients and servers to foundational cloud services. We want those people to love using Async Rust. For that to happen, we need to move Async Rust beyond the "MVP" state it's in today and make it accessible to everyone.
+
+We are launching a collaborative effort to build a shared [vision document][vd] for Async Rust. **Our goal is to engage the entire community in a collective act of the imagination: how can we make the end-to-end experience of using Async I/O not only a pragmatic choice, but a joyful one?**
+
+If you'd like to be a part of that, good news: we need your help! As one of the first steps, we want to encourage people to write about their experiences using Async Rust. Read on for more details!
+
+### Involving the whole community
+
+Building a truly great experience for Async Rust takes more than just `async fn`. We need to think about the end-to-end experience -- everything from what happens when people first learn Async Rust, to what happens as they are building their application, to what happens after deployment.
+
+The challenge for us is to drive that holistic product experience while retaining the loosely coupled, exploration-oriented ecosystem we have built. Async Rust was intentionally designed not to have a "one size fits all" mindset, and we don't want to change that. But we need a forum where we can coordinate on the overall experience we are all shooting for. This is where the [vision doc][vd] comes in.
+
+### To know where you're going, you have to know where you are
+
+Our first goal for the [vision document][vd] is to describe what it's like to use Async Rust today. We've created a [cast of characters][cc] that represent the different backgrounds and experiences that people bring to bear when learning and using Async Rust:
+
+* **Alan**, the full-stack developer working across many companies;
+* **Grace**, the principal engineer hacking on a data storage service;
+* **Niklaus**, the developer building generic Rust libraries and frameworks;
+* **Barbara**, the embedded developer doing networking.
```

Comment:
I've been debating about the characters-- I don't think we want more than 4 characters, it's already kind of hard to remember them. The question to ask yourself is "does this additional character allow me to write a story I couldn't write before"? 

Still, one thing I've observed is that people get distracted by the specifics sometimes, and think that it doesn't apply to them. We've also got an explicit ask later on to tell us if you don't see yourself in the characters.

At minimum, I think it's a good idea to inject some uncertainty into the list, to let people know it's still in flux.

Reply:
(I'll think about it and consider how I want to change it.)

Reply:
I appreciate that one doesn't want to go too granular on personas, but even "embedded" is such a huge field. There's bare-metal, which means close interaction with the hardware facilities, there's a huge class of OSes that bring their own (often task-based) facilities, and then there's also just very small, memory constrained gateway systems that run embedded linux, but do have all the common facilities.

It's not unusual to have ~10-12 personas on projects of such scope.

Reply:
Yeah, I think it might be a bit early in the process to even list these personas right now, unless of course you are interested in a ton of feedback about how they are inadequate ;) (not because I think they are bad but because personas can often be a honeypot for feedback)

If you want to keep them I would encourage you to use language that doesn't make them sound like the full set and actively seek feedback on them.

Reply:
Also, nit but you would probably win a bit by diversifying the names a bit.

Reply:
> I appreciate that one doesn't want to go too granular on personas, but even "embedded" is such a huge field. There's bare-metal, which means close interaction with the hardware facilities, there's a huge class of OSes that bring their own (often task-based) facilities, and then there's also just very small, memory constrained gateway systems that run embedded linux, but do have all the common facilities.

We've been discussing this a bit on Zulip. I think the idea here is to align with the activities of the Embedded WG (which focuses on bare-metal to a large degree). I think we can spin a good story highlighting the high degree of inherent asyncness of such systems and why having good async support in Rust applicable to such systems is a true game changer in the industry. I do agree that the title "Embedded" or even "Embedded networking" is a bit arbitrary, althought I couldn't think of better name. Writing the above paragraph actually made me think maybe we should go "Bare-metal Embedded" or something like that instead. 😅

Reply:
@therealprof i think it would actually be cool to mention that it's a collab with the Embedded WG directly- i don't know if the other personas are based on collabs with other teams/WGs but that seems like it would be 1) just awesome to share because cross team collab is great and 2) would really help reduce the "infinite expansion" potential issue.. if the personas are based on internal team(s) that async is actively collabbing with then it helps focus the audience and prevent it from getting too large or from requiring 100% coverage

Reply:
i think companies like to call using internal teams as audiences "dog fooding" ;) but i think it's a great general purpose technique and think it would work well here

Reply:
I just pushed some edits. I've reworked this section to talk about assembling a cast of characters but not list them all; I now only introduce Grace, since it's her story I'm using as an example. I've also been reworking the overall descriptions of the characters, those comments and thoughts are in the `#wg-async-foundations` stream on Zulip.

(The names, fwiw, comes from famous programming language designers (Alan Turing, Grace Hopper, Barbara Liskov, Niklaus Wirth). I tried a bunch of different ways to derive names, e.g., looking at the most popular names in different parts of the world and so forth, but this seemed like a decent way to ground them.)